torch/csrc/jit/serialization/import.cpp

```diff
@@ -144,35 +145,48 @@ IValue ScriptModuleDeserializer::readArchive(const std::string& archive_name) {
   // Method.run() and graph executor, etc.
   // For bytecode import we need to decouple these dependencies.
   auto obj_loader = [&](at::StrongTypePtr type, IValue input) {
-    auto cls = type.type_->expect<at::ClassType>();
-    auto qn = cls->name();
-    size_t n = cls->numAttributes();
-    if (checkHasValidSetGetState(cls)) {
-      auto obj = c10::ivalue::Object::create(type, n);
-      // XXX: Do not optimize __setstate__, so that we don't try to
-      // specialize the class before it is initialized.
-      GraphOptimizerEnabledGuard guard(false);
-      Function& set_state = cls->getMethod("__setstate__");
-      // since we are in the middle of unpickling we might still have lists and
-      // dicts that do not have accurate tags (e.g. they report they are
-      // List[Any]). But we need to run __setstate__ which will check the input
-      // type and may access the tags. Since setstate has a known input type, we
-      // can correctly restore the tags now by apply the input type of set_state
-      // to the state object being passed.
-      // TODO: Remove once [serialization type tags] is landed
-      restoreAccurateTypeTags(
-          input, set_state.getSchema().arguments().at(1).type());
-      set_state({obj, input});
-      postSetStateValidate(obj);
-      return obj;
-    } else {
-      auto dict = std::move(input).toGenericDict();
-      auto obj = c10::ivalue::Object::create(type, n);
-      for (size_t i = 0; i < n; ++i) {
-        obj->setSlot(i, dict.at(cls->getAttributeName(i)));
+    if (auto enum_type = type.type_->cast<EnumType>()) {
```

Comment:
I have context on the picker, but I didn't review the Enum changes so I don't know how they are implemented. That said,  my gut instinct is something if fishy about how Enums are/are not treated as classes here and go through a different depickling process: if they are classes, what if someone defines setstate, looks like that gets skipped? If they are not, why do they go through the object loader? If the pickler.cpp was edited for serialization, then unpickle.cpp should be the place where we do the deserialization.

Reply:
Thanks for the suggestion. I also had some discomfort with changing objLoader.

Just to make sure I understand it correctly, do you suggest we add a special case here:https://github.com/pytorch/pytorch/blob/4dcd2781b8f3de59bb1a5d6e337faad7f68c8e7b/torch/csrc/jit/serialization/unpickler.cpp#L594

to handle EnumType so that objLoader is still pure?

Reply:
yep

Reply:
Done. Please take another look.